### PR TITLE
[xabuild.exe] another fix for multi-process file access

### PR DESCRIPTION
**Pull request by jonathanpeppers**
Context: d82c97c
Context: https://devdiv.visualstudio.com/DevDiv/_build/index?buildId=1550431

For a while, my previous changes to `xabuild.exe` seemed to be working.

It still failed sometimes after reviewing past builds on VSTS:
- `RunXABuildInParallel` test failed randomly with non-zero exit code
when running `xabuild /version`
- Any test in `Xamarin.Android.Build.Tests`

The exception is something like:

    Unhandled Exception: System.IO.IOException: The process cannot access the file 'bin\Debug\lib\xamarin.android\xbuild-frameworks\.__sys_links.txt' because it is being used by another process.

I think one more change might finally fix this. Before this change,
`xabuild.exe` was opening the `.__sys_links.txt` every time. It should
first check and see if the symbolic links already exist before opening
the file.

I also added a `Console.WriteLine` to list every file link, and I soon
saw something odd from the logs on VSTS:
```
[xabuild] creating symbolic link 'E:\A\_work\2\s\bin\Debug\lib\xamarin.android\xbuild-frameworks\MonoAndroid'
    -> 'C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\MonoAndroid'
```

I remoted into the machine, and there was a symbolic link leftover from
`setup-windows.exe` (I believe) that was pointing from
`C:\Program Files (x86)\Reference Assemblies` to `xamarin-android` build
output somewhere!

Changes (summary):
- Check if any symbolic links are needed before opening
`.__sys_links.txt`
- Added a special case for a `MonoAndroid` directory on a VSTS build
agent (I will login and remove this directory on the agent later)
- Added a `Console.WriteLine` for when file links are created.
- Additionally, we should be using `File.AppendText` instead of
`File.CreateText` as we don't want to *ever* completely overwrite
`.__sys_links.txt`
- Refactored `SymbolicLink` so it more clearly calls `Directory.Exists`
one last time before deciding the symbolic link creation was a failure.
Also moved the Windows-specific code to its own method.

### CODE REVIEW

**Comment by jonathanpeppers** [comment]
Fix didn't help on Windows 😞 

Going to try something else

**Comment by dellis1972** [comment]
build

**Comment by jonathanpeppers** [comment]
Something odd is going on, this isn't happening for me locally:
```
2018-04-05T14:08:41.2112225Z [xabuild] creating symbolic link 'E:\A\_work\2\s\bin\Debug\lib\xamarin.android\xbuild-frameworks\.NETCore' -> 'C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETCore'
2018-04-05T14:08:41.2113604Z [xabuild] creating symbolic link 'E:\A\_work\2\s\bin\Debug\lib\xamarin.android\xbuild-frameworks\.NETFramework' -> 'C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework'
2018-04-05T14:08:41.2114824Z [xabuild] creating symbolic link 'E:\A\_work\2\s\bin\Debug\lib\xamarin.android\xbuild-frameworks\.NETPortable' -> 'C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETPortable'
2018-04-05T14:08:41.2116019Z [xabuild] creating symbolic link 'E:\A\_work\2\s\bin\Debug\lib\xamarin.android\xbuild-frameworks\MonoAndroid' -> 'C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\MonoAndroid'
2018-04-05T14:08:41.2117142Z [xabuild] creating symbolic link 'E:\A\_work\2\s\bin\Debug\lib\xamarin.android\xbuild-frameworks\v3.0' -> 'C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\v3.0'
2018-04-05T14:08:41.2118239Z [xabuild] creating symbolic link 'E:\A\_work\2\s\bin\Debug\lib\xamarin.android\xbuild-frameworks\v3.5' -> 'C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\v3.5'
```
A link to a `MonoAndroid` directory?

**Comment by jonathanpeppers** [comment]
Finally! @dellis1972 this one is good to go if the Jenkins builds pass.

This test has been failing on Windows: `AndroidUpdateResourcesTest.ReportAaptErrorsInOriginalFileName`, but I will fix that test in a separate PR.

**Comment by dellis1972** [comment]
@jonathanpeppers weird, the api compatibility check failed on macOS+xbuild... it doesn't look related though so I am tempted to merge this.

**Comment by jonathanpeppers** [comment]
They were talking about System.Memory api changes with mono bumps yesterday. Is it because mono got upgraded on the build machine and this PR is a few commits behind?

Otherwise xabuild.exe shouldn’t run on that build anyway.

**Comment by jonathanpeppers** [comment]
I just rebased, hopefully everything will be green.

File: tools/xabuild/XABuild.cs [hunk 1]
```diff
@@ -21,15 +22,26 @@ static int Main ()
...
-							return 1;
+				//Symbolic links to .NETFramework and .NETPortable directory
+				var systemDirectories = Directory.EnumerateDirectories (paths.SystemProfiles)
+					.Where (d => Path.GetFileName (d) != "MonoAndroid") //NOTE: this happened on one of our VSTS build agents
+					.ToArray ();
+				var symbolicLinks = systemDirectories
+					.Select (d => Path.Combine (paths.FrameworksDirectory, Path.GetFileName (d)))
+					.ToArray ();
+
+				if (symbolicLinks.Any (d => !Directory.Exists (d))) {
```
Comment (jonpryor) [inline tools/xabuild/XABuild.cs]: This doesn’t make sense to me.

Imagine two `xabuild.exe` processes started at the “same” time, both hitting the above condition at the “same” time, with at least one directory not existing, causing this `if` to be entered.

What I believe would happen is that they *both* would execute the following `for` loop, creating the symbolic links. The “first” one would “win”, creating the symbolic links and the `.__sys_links.txt` file.

The “second” one, I *suspect* (but have not tested), would later execute `OpenSysLinksFile()` — *clearing* the file contents! — then *exit* as`SymbolicLink.Create()` returns `false` (assuming that `CreateSymbolicLink()` fails if the target file already exists; the docs don’t specify).

The end state being that the 2nd `xabuild.exe` does *nothing*, *and* `.__sys_links.txt` is cleared.

This sounds like a bad combination. :-(

Comment (jonathanpeppers) [reply tools/xabuild/XABuild.cs]: Whoops, this needs to use `File.AppendText` instead of `File.CreateText`.

Comment (jonpryor) [reply tools/xabuild/XABuild.cs]: Even with `File.AppendText()`, that just introduces duplication within `.__sys_links.txt`. Which might be fine, actually...

That doesn't address the 2nd `xabuild.exe` exiting once it fails to create a symlink because the symlink already exists.

Comment (jonathanpeppers) [reply tools/xabuild/XABuild.cs]: `File.AppendText` should throw an `IOException` if two processes try to call it at the same time:

image

`SymbolicLink` has its own `Directory.Exists` check: https://github.com/xamarin/xamarin-android/blob/3f7eca469382a29e67ec799fce1e62510d2291f1/tools/xabuild/SymbolicLink.cs#L12

Do you think that is enough? I have been running this test after deleting the symlinks to test this: https://github.com/xamarin/xamarin-android/blob/0dc96b02d9392adc136dc92469b00df7cb999a41/src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/BuildTest.cs#L2422

Comment (jonpryor) [reply tools/xabuild/XABuild.cs]: > Do you think [the `Directory.Exists()` check] is enough?

No. That's the fun of multithreading! There's no cross-process locking going on, so I can imagine the `Directory.Exists(source)` check returning false (it doesn't exist), then the process is suspended, the *second* process comes in and actually creates the symlink, and thus it exists before the first process creates the symlink, which fails.

...except we *do* have cross-process locking on Windows, because of the file access...

...except that *mono* on Linux & macOS does *not* support file sharing across process boundaries:


```csharp
using System;
using System.IO;

class App {
	public static void Main ()
	{
		var path = "foo.txt";
		var x = File.AppendText (path);
		Console.ReadLine ();
	}
}
```

Run the above app in two different shell windows within the same directory, and no exception is thrown.

Additionally, we *want* to use `xabuild.exe` for Linux & macOS builds, as part of moving away from xbuild...

TL;DR: I believe that instead of a `Directory.Exists()` check *before* the symlink creation, we should instead (additionally?) do the `Directory.Exists()` check when the symlink creation *fails*, and if `Directory.Exists()` is true, then "ignore" the symlink creation failure, because the symlink *does* exist.

Comment (jonathanpeppers) [reply tools/xabuild/XABuild.cs]: Ok I see now.

Maybe a small test project is worth it? `xabuild-Tests`? Assuming the entire test run only takes a second or something.

Comment (jonpryor) [reply tools/xabuild/XABuild.cs]: I'm not opposed to unit tests, though I'm not entirely sure i see what they'd be testing/how you'd be testing it...
